srtcore/core.cpp

```diff
@@ -9834,3 +10028,818 @@ bool CUDT::runAcceptHook(CUDT *acore, const sockaddr* peer, const CHandShake& hs
+        const SRT_MSGCTRL& in, // MSGCTRL read from the data-providing socket
+        SRT_SOCKGROUPDATA* out_grpdata, // grpdata as passed in MSGCTRL
+        size_t out_grpdata_size) // grpdata_size as passed in MSGCTRL
+{
+    w_out = in;
+
+    // User did not wish to read the group data at all.
+    if (!out_grpdata)
+    {
+        w_out.grpdata = NULL;
+        w_out.grpdata_size = 0;
+        return;
+    }
+
+    int st = getGroupData((out_grpdata), (&out_grpdata_size));
+    w_out.grpdata_size = out_grpdata_size;
+    // On error, rewrite NULL.
+    w_out.grpdata = st == 0 ? out_grpdata : NULL;
+}
+
+struct FLookupSocketWithEvent
+{
+    CUDTUnited* glob;
+    int evtype;
+    FLookupSocketWithEvent(CUDTUnited* g, int event_type): glob(g), evtype(event_type) {}
+
+    typedef CUDTSocket* result_type;
+
+    pair<CUDTSocket*, bool> operator()(const pair<SRTSOCKET, int>& es)
+    {
+        CUDTSocket* so = NULL;
+        if ( (es.second & evtype) == 0)
+            return make_pair(so, false);
+
+        so = glob->locateSocket(es.first, glob->ERH_RETURN);
+        return make_pair(so, !!so);
+    }
+};
+
+void CUDTGroup::updateReadState(SRTSOCKET /* not sure if needed */, int32_t sequence)
+{
+    bool ready = false;
+    CGuard lg (m_GroupLock);
+    int seqdiff = 0;
+
+    if (m_RcvBaseSeqNo == -1)
+    {
+        // One socket reported readiness, while no reading operation
+        // has ever been done. Whatever the sequence number is, it will
+        // be taken as a good deal and reading will be accepted.
+        ready = true;
+    }
+    else if ( (seqdiff = CSeqNo::seqcmp(sequence, m_RcvBaseSeqNo)) > 0 )
+    {
+        // Case diff == 1: The very next. Surely read-ready.
+
+        // Case diff > 1:
+        // We have an ahead packet. There's one strict condition in which
+        // we may believe it needs to be delivered - when KANGAROO->HORSE
+        // transition is allowed. Stating that the time calculation is done
+        // exactly the same way on every link in the redundancy group, when
+        // it came to a situation that a packet from one link is ready for
+        // extraction while it has jumped over some packet, it has surely
+        // happened due to TLPKTDROP, and if it happened on at least one link,
+        // we surely don't have this packet ready on any other link.
+
+        // This might prove not exactly true, especially when at the moment
+        // when this happens another link may surprisinly receive this lacking
+        // packet, so the situation gets suddenly repaired after this function
+        // is called, the only result of it would be that it will really get
+        // the very next sequence, even though this function doesn't know it
+        // yet, but surely in both cases the situation is the same: the medium
+        // is ready for reading, no matter what packet will turn out to be
+        // returned when reading is done.
+
+        ready = true;
+    }
```

Comment:
Will the body of both conditions contain more logic in further PRs? Right now they share the one and only statement.

Reply:
No, but mind that these two conditional expressions can't be bound into one anyway. And this comment is also important.